### PR TITLE
Trying to augment argument set with Evaluator and Expression context

### PR DESCRIPTION
**Pull request by CTrando**

### CODE REVIEW

File: src/Analysis/Ast/Impl/Values/Collections/PythonDictionary.cs [hunk 1]
```diff
@@ -62,7 +62,7 @@ public IReadOnlyList<IPythonCollection> Items
             _contents.TryGetValue(key, out var value) ? value : UnknownType;
 
         public override IPythonIterator GetIterator() =>
-            Call(@"iterkeys", ArgumentSet.Empty) as IPythonIterator ?? new EmptyIterator(Type.DeclaringModule.Interpreter.UnknownType);
+            Call(@"iterkeys", null) as IPythonIterator ?? new EmptyIterator(Type.DeclaringModule.Interpreter.UnknownType);
```
Comment (jakebailey) [inline src/Analysis/Ast/Impl/Values/Collections/PythonDictionary.cs]: null?

Comment (CTrando) [reply src/Analysis/Ast/Impl/Values/Collections/PythonDictionary.cs]: ah my bad should be ArgumentSet.EmptyNoContext

File: src/Analysis/Ast/Impl/Types/ArgumentSet.cs [hunk 2]
```diff
@@ -38,7 +39,7 @@ internal sealed class ArgumentSet : IArgumentSet {
         private readonly DictArg _dictArgument;
         private bool _evaluated;
 
-        public static IArgumentSet Empty = new ArgumentSet();
+        public static IArgumentSet EmptyNoContext = new ArgumentSet();
```
Comment (jakebailey) [inline src/Analysis/Ast/Impl/Types/ArgumentSet.cs]: Perhaps just `WithoutContext`, such that it's used as `ArgumentSet.WithoutContext`.

File: src/Analysis/Ast/Impl/Types/ArgumentSet.cs [hunk 3]
```diff
@@ -49,19 +50,38 @@ internal sealed class ArgumentSet : IArgumentSet {
...
+        /// </summary>
+        /// <param name="expr">Expression associated with argument set.</param>
+        /// <param name="eval">Evaluator for the expression involving the argument set.</param>
+        /// <returns></returns>
+        public static ArgumentSet Empty(Expression expr, IExpressionEvaluator eval) {
+            return new ArgumentSet(new List<IMember>(), expr, eval);
         }
+        
+        /// <summary>
+        /// Creates a set of arguments for a constructor call, an indexing call, etc.
```
Comment (jakebailey) [inline src/Analysis/Ast/Impl/Types/ArgumentSet.cs]: Just "a call".

File: src/Analysis/Ast/Impl/Types/ArgumentSet.cs (hunk 2 above)
Comment (CTrando) [reply src/Analysis/Ast/Impl/Types/ArgumentSet.cs]: Done!

File: src/Analysis/Ast/Impl/Types/PythonFunctionOverload.cs [hunk 4]
```diff
@@ -167,7 +167,7 @@ public PythonFunctionOverload(IPythonClassMember cm, FunctionDefinition fd, Loca
             }
 
             if (typeArgs != null) {
-                var specificReturnValue = returnClassType.CreateSpecificType(new ArgumentSet(typeArgs));
+                var specificReturnValue = returnClassType.CreateSpecificType(new ArgumentSet(typeArgs, args.Expression, args.Eval));
```
Comment (jakebailey) [inline src/Analysis/Ast/Impl/Types/PythonFunctionOverload.cs]: If you run MethodFromBaseClass, this fails here, seemingly because `args` is null.

Comment (jakebailey) [reply src/Analysis/Ast/Impl/Types/PythonFunctionOverload.cs]: GetReturnDocumentation calls with `null` args explicitly, which triggers the if statement above. I think it's probably on purpose, so maybe just `args?` here is good enough.

Comment (CTrando) [reply src/Analysis/Ast/Impl/Types/PythonFunctionOverload.cs]: Thanks! Fixed.

**Pull request merged**